proposals/2022/07-01_nested_api_objects.md

```diff
@@ -0,0 +1,244 @@
+| [library.Hook](https://github.com/go-vela/types/blob/master/library/hook.go#L11-L29) | `repo_id`, `build_id` |
+| [library.Log](https://github.com/go-vela/types/blob/master/library/log.go#L14-L25) | `repo_id`, `build_id`, `service_id`, `step_id` |
+| [library.Pipeline](https://github.com/go-vela/types/blob/master/library/pipeline.go#L11-L31) | `repo_id` |
+| [library.Repo](https://github.com/go-vela/types/blob/master/library/repo.go#L11-L38) | `user_id` |
+| [library.Service](https://github.com/go-vela/types/blob/master/library/service.go#L16-L35) | `repo_id`, `build_id` |
+| [library.Step](https://github.com/go-vela/types/blob/master/library/step.go#L16-L36) | `repo_id`, `build_id` |
+
+> NOTE: To provide credit where it's due, this pattern is being adopted from an existing, prominent API (GitHub).
+>
+> To see an example of how this may look in a real world scenario, you can reference their [docs on getting a repo](https://docs.github.com/en/rest/repos/repos#get-a-repository).
+>
+> The `owner` field from the GitHub docs displays the behavior being referred to.
+
+<!--
+Provide your description here.
+-->
+
+**Please briefly answer the following questions:**
+
+**1. Why is this required?**
+
+<!-- Answer here -->
+
+This functionality should not be considered required.
+
+Instead, this proposal aims to enhance/augment the existing API implementation.
+
+**2. If this is a redesign or refactor, what issues exist in the current implementation?**
+
+<!-- Answer here -->
+
+Today, the API returns pieces of information that aren't valuable without directly querying the database.
+
+This functionality would improve the experience for both administrators and end-users.
+
+i.e. an end-user is able to answer the question "Who is the owner of a repo in Vela?"
+
+**3. Are there any other workarounds, and if so, what are the drawbacks?**
+
+<!-- Answer here -->
+
+N/A
+
+**4. Are there any related issues? Please provide them below if any exist.**
+
+<!-- Answer here -->
+
+https://github.com/go-vela/community/issues/69
+
+## Design
+
+<!--
+This section is intended to explain the solution design for the proposal.
+
+NOTE: If there are no current plans for a solution, please leave this section blank.
+-->
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* new/updated endpoints or url paths
+* new/updated configuration variables (environment, flags, files, etc.)
+* performance and user experience tradeoffs
+* security concerns or assumptions
+* examples or (pseudo) code snippets
+
+<!-- Answer here -->
+
+As described, the idea is to no longer return an ID field for related objects from the API.
+
+Using the above endpoint (`GET /api/v1/repo/:org/:repo`), the result would replace the `user_id` field with an `owner` field:
+
+```diff
+{
+  "id": 1,
+- "user_id": 1,
++ "owner": {
++   "id": 1,
++   "name": "OctoKitty",
++   "favorites": ["github/octocat"],
```

Comment:
I don't think favorites should be exposed in most cases. It seems to me that that's a user preference that the UI (or CLI) should be able to request for the current user, bit nobody/nothing else needs it.

What purpose does exposing a user's favorites to a build (in env vars) or in these API calls have? Is there a use case I haven't considered?

Reply:
I don't think we necessarily meant to propose that the `owner` field has to have the `favorites` displayed.

More so, I think we simply took the [existing user API definition](https://github.com/go-vela/types/blob/master/library/user.go#L13-L25):

```go
// User is the library representation of a user.
//
// swagger:model User
type User struct {
	ID           *int64    `json:"id,omitempty"`
	Name         *string   `json:"name,omitempty"`
	RefreshToken *string   `json:"-"`
	Token        *string   `json:"-"`
	Hash         *string   `json:"-"`
	Favorites    *[]string `json:"favorites,omitempty"`
	Active       *bool     `json:"active,omitempty"`
	Admin        *bool     `json:"admin,omitempty"`
}
```

And removed the fields that can't be returned (`-` in JSON field tags).

Regarding implementation, we can certainly ensure that `favorites` is not returned to minimize payload size.